Add support for specifying negated queues

bdragon opened the pull request:
## What this change does

Augments the syntax used to specify worker queues to support negation.

Sample usage:

```
$ QUEUES=!priority,* rake resque:work
```

## Motivation

We would like to configure one worker to handle a specific queue, say `priority`, while configuring other workers to work off of all queues (splat) *except* for the `priority` queue. Our desired Procfile looks like this:

```
priority: env TERM_CHILD=1 QUEUES=priority bundle exec rake resque:work
worker: env TERM_CHILD=1 COUNT=3 QUEUES=!priority,* bundle exec rake resque:work
```

This is currently not possible without enumerating a whitelist of queues that workers should work off of, which is not an option for us as we would like to continue to have workers pick up new queues as they are added.

jeremy commented on lib/resque/worker.rb:
```diff
@@ -178,9 +178,15 @@ def prepare
-      @queues = queues.map { |queue| queue.to_s.strip }
-      @has_dynamic_queues = WILDCARDS.any? {|char| @queues.join.include?(char) }
+      queues = (ENV["QUEUES"] || ENV['QUEUE']).to_s.split(',') if queues.empty?
+      queues.map! { |queue| queue.to_s.strip }
```
Mutates the `queues` argument.

bryandragon replied:
Thanks—updated.

jeremy commented on README.markdown:
```diff
@@ -354,6 +354,19 @@ queues created on the fly, you can use a splat:
+
+    $ QUEUE=*,!file_*
+
+
```
Worth nothing that order doesn't matter for skipped queues?

bryandragon replied:
Good call. I updated accordingly.

jrochkind commented:
This feature would be super useful to me. 

Any chance of it landing in resque? Status? (Is resque still maintained?)

mbiffara commented:
Is there a chance we can merge this? It'd be very useful

iloveitaly commented on README.markdown:
```diff
@@ -354,6 +354,23 @@ queues created on the fly, you can use a splat:
+matter, so `QUEUE=*,!file_*` and `QUEUE=!file_*,*` will have the same
+effect.
+
+
```
Thanks for the clear documentation!

iloveitaly commented on lib/resque/worker.rb:
```diff
@@ -178,9 +178,15 @@ def prepare
+      @skip_queues.map! { |queue| queue[1..-1] }
+
+      @has_dynamic_queues =
+        @skip_queues.any? || WILDCARDS.any? { |char| @queues.join.include?(char) }
```
I'm not following the `@skip_queues.any?` condition here, could you explain? Why is this required?

bdragon replied:
The behavior of `#queues` is dependent on the value of `@has_dynamic_queues`: if it's true, the method returns the result of filtering `@queues` with`#glob_match`; if it's false, the method returns `@queues` directly. Since `#glob_match` will cause skipped queues to be filtered out, we want to make sure it's called if we have  `@skip_queues.any?`.

https://github.com/resque/resque/blob/92a4564fac9af56ad0b1b2a37ad51dbea5c2bf36/lib/resque/worker.rb#L202-L215

It's like wildcards: just as we don't want to return `my-queue-*` from `#queues`, we don't want to return `!my-queue`, but rather the result of filtering `@queues` with the pattern.

iloveitaly replied:
Got it, thanks for explaining!

iloveitaly reviewed (changes requested):
Thanks for the contribution! I know this is an old PR, but I'd still love to get it in. Just one bit I'd like some clarification on before merging.

Also, could you rebase on master so CI runs?

@mbiffara @jrochkind if either of you want to help out with the last details on this PR, feel free!

bdragon commented:
@iloveitaly This is ready to go. I rebased against master and I believe I addressed your question [here](https://github.com/resque/resque/pull/1617#discussion_r734175333). Let me know if there's anything else!

ksweetie commented:
@iloveitaly I'd love to get this one merged—is there anything else we can do to help get it across the line? Thanks for your help!

Murphydbuffalo commented:
@iloveitaly just bumping this thread, do you think you'll be able to merge this sometime in the near future? Thanks! Happy to help out if that would be useful

iloveitaly commented on lib/resque/worker.rb:
```diff
@@ -180,9 +180,15 @@ def prepare
+      @skip_queues, @queues = queues.partition { |queue| queue.start_with?('!') }
+      @skip_queues.map! { |queue| queue[1..-1] }
+
+      @has_dynamic_queues =
```
```suggestion
      # The behavior of `queues` is dependent on the value of `@has_dynamic_queues: if it's true, the method returns the result of filtering @queues with `glob_match`
      # if it's false, the method returns @queues directly. Since `glob_match` will cause skipped queues to be filtered out, we want to make sure it's called if we have @skip_queues.any?
      @has_dynamic_queues =
```

iloveitaly commented:
Sorry for the delay! Thanks for this contribution
